### PR TITLE
[Transform] Reduce indexes to query based on checkpoints

### PR DESCRIPTION
**Pull request by hendrikmuhs**
Continuous transform reduce the amount of data to query for by detecting what has been changed
since the last checkpoint. This information is than used to inject queries that narrow the
scope. The query is send to all configured indices. This change reduces the indexes to call
using checkpoint information. This reduces not only the number of network calls, but also
reduces the probability of a failure, which is more likely to happen in large heterogeneous
clusters (hot/warm/cold architecture).

### CODE REVIEW

**Comment by hendrikmuhs** [comment]
TL/DR

I think I need to explain the idea for part 2, reducing calls for `pivot`:

The change collector already collects meta information about the source indexes, e.g. it gets `max(@timestamp)`. I want to keep this information and store it in the state document. Imagine we keep the last 5 checkpoints (whether its 5 or something higher needs to be tested):

```
_transform.cp.42.@timestamp.max: 999
_transform.cp.43.@timestamp.max: 1022
_transform.cp.44.@timestamp.max: 1080
_transform.cp.45.@timestamp.max: 1129
_transform.cp.46.@timestamp.max: 1535
```

Assume for checkpoint `47` we detected in the change collector that `@timestamp` has a range `min: 1092`, `max: 1693`. The above lookup table tells us, we can skip every index that hasn't changed _after_ checkpoint `44`. Therefore we call `getChangedIndices(44, 47)` which gives us the indices with changes.

In a nutshell the checkpoint for `44` could look like this:

```
index-0001: 1220,1220,10135,1015,10111
index-0002: 305,15134,13531,1352,15533
...
index-0099: 10, 10, 10, 10, 10
index-0101: 20, 30, 30, 30, 30
index-0102: 40, 40, 40, 40, 40
```

The checkpoint of `47` might be:

```
index-0001: 1220,1220,10135,1015,10111
index-0002: 305,15134,13531,1352,15533
...
index-0099: 10, 10, 10, 10, 10
index-0101: 23, 32, 42, 32, 60
index-0102: 40, 50, 70, 60, 50
index-0103: 10, 10, 10, 10, 10
```

If we diff the 2, we can remove [`index-0001`, ... `index-0099`], only these 3 indexes are of interest:

```
index-0101: 23, 32, 42, 32, 60
index-0102: 40, 50, 70, 60, 50
index-0103: 10, 10, 10, 10, 10
```

We therefore can resolve the pattern `index-*` to only call the shards of those 3 indexes instead of calling all of them.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 1]
```diff
@@ -1091,32 +1091,34 @@ protected QueryBuilder buildFilterQuery() {
...
-    protected SearchRequest buildSearchRequest() {
+    protected Tuple<String, SearchRequest> buildSearchRequest() {
         assert nextCheckpoint != null;
 
-        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().runtimeMappings(getConfig().getSource().getRuntimeMappings());
         switch (runState) {
             case APPLY_RESULTS:
-                buildUpdateQuery(sourceBuilder);
-                break;
+                return new Tuple<>("apply_results", buildQueryToUpdateDestinationIndex());
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1099]: I really like this. Naming them like this makes debugging in the future much nicer!!!

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 2]
```diff
@@ -433,24 +445,26 @@ private void injectPointInTimeIfNeeded(SearchRequest searchRequest, ActionListen
...
-                listener.onResponse(searchRequest);
+                listener.onResponse(namedSearchRequest);
             })
         );
     }
 
-    private void doSearch(SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
-        logger.trace("searchRequest: {}", searchRequest);
+    private void doSearch(Tuple<String, SearchRequest> namedSearchRequest, ActionListener<SearchResponse> listener) {
+        logger.trace("searchRequest: {}", namedSearchRequest.v2());
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:454]: This may serialize `namedSearchRequest.v2()` string even when trace is not turned on correct? That could be expensive.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 3]
```diff
@@ -433,24 +445,26 @@ private void injectPointInTimeIfNeeded(SearchRequest searchRequest, ActionListen
...
             client,
             SearchAction.INSTANCE,
-            searchRequest,
+            namedSearchRequest.v2(),
             ActionListener.wrap(response -> {
                 // did the pit change?
                 if (response.pointInTimeId() != null && (pit == null || response.pointInTimeId() != pit.getEncodedId())) {
-                    pit = new PointInTimeBuilder(response.pointInTimeId()).setKeepAlive(PIT_KEEP_ALIVE);
+                    namedPits.put(namedSearchRequest.v1(), new PointInTimeBuilder(response.pointInTimeId()).setKeepAlive(PIT_KEEP_ALIVE));
                     logger.trace("point in time handle has changed");
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:468]:
```suggestion
                    logger.trace("point in time handle has changed; request [{}]", namedSearchRequest.v1());
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 4]
```diff
@@ -462,14 +476,14 @@ private void doSearch(SearchRequest searchRequest, ActionListener<SearchResponse
                 Throwable unwrappedException = ExceptionsHelper.findSearchExceptionRootCause(e);
                 if (unwrappedException instanceof SearchContextMissingException) {
                     logger.warn(new ParameterizedMessage("[{}] Search context missing, falling back to normal search.", getJobId()), e);
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:478]:
```suggestion
                    logger.warn(new ParameterizedMessage("[{}] Search context missing, falling back to normal search; request [{}]", getJobId(), namedSearchRequest.v1()), e);
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 5]
```diff
@@ -405,11 +416,12 @@ private void injectPointInTimeIfNeeded(SearchRequest searchRequest, ActionListen
...
             ActionListener.wrap(response -> {
-                pit = new PointInTimeBuilder(response.getPointInTimeId()).setKeepAlive(PIT_KEEP_ALIVE);
-                searchRequest.source().pointInTimeBuilder(pit);
+                PointInTimeBuilder newPit = new PointInTimeBuilder(response.getPointInTimeId()).setKeepAlive(PIT_KEEP_ALIVE);
+                namedPits.put(namedSearchRequest.v1(), newPit);
+                searchRequest.source().pointInTimeBuilder(newPit);
                 pitCheckpoint = getNextCheckpoint().getCheckpoint();
-                logger.trace("[{}] using pit search context with id [{}]", getJobId(), pit.getEncodedId());
-                listener.onResponse(searchRequest);
+                logger.trace("[{}] using pit search context with id [{}]", getJobId(), newPit.getEncodedId());
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:423]:
```suggestion
                logger.trace("[{}] using pit search context with id [{}]; request [{}]", getJobId(), newPit.getEncodedId(), namedSearchRequest.v1());
```

**Review by benwtrent — APPROVED** [review]
I think the idea is sound and will definitely help resiliency. I don't think it will help much in the performance front.

I would like to add the `name` for the PIT and searchrequest to the logging messages in some fashion. 

Additionally, I am slightly concerned that building the checkpoints themselves will still fail as they have to get the stats from all indices. Though, I am not 100% sure how that index stats request differs from searching them.
